packages/material-ui/src/Hidden/HiddenCss.js

```diff
@@ -1,42 +1,85 @@
 import * as React from 'react';
+import clsx from 'clsx';
 import PropTypes from 'prop-types';
+import { unstable_composeClasses as composeClasses } from '@material-ui/unstyled';
 import capitalize from '../utils/capitalize';
-import withStyles from '../styles/withStyles';
+import experimentalStyled from '../styles/experimentalStyled';
+import useThemeProps from '../styles/useThemeProps';
 import useTheme from '../styles/useTheme';
+import { getHiddenCssUtilityClass } from './hiddenCssClasses';
 
-const styles = (theme) => {
+const sizes = ['xl', 'lg', 'md', 'sm', 'xs'];
+
+const overridesResolver = (props, styles) => styles.root || {};
+
+const useUtilityClasses = (styleProps) => {
+  const { classes, breakpoints } = styleProps;
+
+  const slots = {
+    root: [
+      'root',
+      ...sizes.map((size) => breakpoints.includes(`${size}Up`) && `${size}Up`),
+      ...sizes.map((size) => breakpoints.includes(`${size}Down`) && `${size}Down`),
+      ...sizes.map(
+        (size) => breakpoints.includes(`only${capitalize(size)}`) && `only${capitalize(size)}`,
+      ),
+    ],
+  };
+
+  return composeClasses(slots, getHiddenCssUtilityClass, classes);
+};
+
+const HiddenCssRoot = experimentalStyled(
+  'div',
+  {},
+  {
+    name: 'PrivateHiddenCss',
+    slot: 'Root',
+    overridesResolver,
+  },
+)(({ theme, styleProps }) => {
   const hidden = {
     display: 'none',
   };
 
-  return theme.breakpoints.keys.reduce((acc, key) => {
-    acc[`only${capitalize(key)}`] = {
-      [theme.breakpoints.only(key)]: hidden,
-    };
-    acc[`${key}Up`] = {
-      [theme.breakpoints.up(key)]: hidden,
-    };
-    acc[`${key}Down`] = {
-      [theme.breakpoints.down(key)]: hidden,
-    };
-
-    return acc;
-  }, {});
-};
+  const styles = theme.breakpoints.keys
```

Comment:
I believe we should base this on the `styleProps.breakpoints` - whatever is there should be hidden.